### ISSUE TITLE
EIP-2930 accessList structure discrepancy

### ISSUE THREAD

**Issue by oocrypticoo** [ethereum/web3.py#2140]
Unable to signTransactions with accessList in Web3.py 5.23.0

**Comment by fselmo** [comment]
@oocrypticoo did you try explicitly setting the `type` for the transaction to `'0x1'` or `1` in the transaction parameters? This is the appropriate transaction type for access list transactions. Access lists can also be signed with type `'0x2'` transactions but those would involve using EIP-1559 fee values (`maxFeePerGas` and `maxPriorityFeePerGas`) instead of the `gasPrice`.

Can you check if this works for you? Set the type to `'0x1'` if using `gasPrice` or `'0x2'` if using 1559 fees, in your transaction params:

```python
tx = {
    'type': '0x1',
    'gasPrice': ...,
    'to': ...,
    'from': ...,
    'value': ...,
    'gas': ...,
    'nonce': ...,
    'accessList': ...,
}
w3.eth.account.signTransaction(tx, PRIVATE_KEY)
```

**Comment by oocrypticoo** [comment]
> @oocrypticoo did you try explicitly setting the `type` for the transaction to `'0x1'` or `1` in the transaction parameters? This is the appropriate transaction type for access list transactions. Access lists can also be signed with type `'0x2'` transactions but those would involve using EIP-1559 fee values (`maxFeePerGas` and `maxPriorityFeePerGas`) instead of the `gasPrice`.
> 
> Can you check if this works for you? Set the type to `'0x1'` if using `gasPrice` or `'0x2'` if using 1559 fees, in your transaction params:
> 
> ```python
> tx = {
>     'type': '0x1',
>     'gasPrice': ...,
>     'to': ...,
>     'from': ...,
>     'value': ...,
>     'gas': ...,
>     'nonce': ...,
>     'accessList': ...,
> }
> w3.eth.account.sign_transaction(tx, PRIVATE_KEY)
> ```
That doesn't work
I'll just paste my tx_details

{
'value': 0,
 'chainId': 1,
 'nonce': 33,
 'from': '0x00000000000000000000000000000',
 'gas': 696969,
 'gasPrice': 148241644394,
 'type': '0x1',
 'accessList': [{'address': '0x5a98fcbea516cf06857215779fd812ca3bef1b32',
   'storageKeys': []},
  {'address': '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2', 'storageKeys': []},
  {'address': '0x454f11d58e27858926d7a4ece8bfea2c33e97b13',
   'storageKeys': ['0x0000000000000000000000000000000000000000000000000000000000000006',
    '0x0000000000000000000000000000000000000000000000000000000000000007',
    '0x0000000000000000000000000000000000000000000000000000000000000008',
    '0x000000000000000000000000000000000000000000000000000000000000000c']},
  {'address': '0xc558f600b34a5f69dd2f0d06cb8a88d829b7420a',
   'storageKeys': ['0x0000000000000000000000000000000000000000000000000000000000000006',
    '0x0000000000000000000000000000000000000000000000000000000000000007',
    '0x0000000000000000000000000000000000000000000000000000000000000008',
    '0x000000000000000000000000000000000000000000000000000000000000000c']}],
 'to': '0x0131CCeB31F7041A60CEC39A89AE832672e17054',
 'data': '0x27653ec6000000000000000000000000000000000000000000000000029f7e71dc3513800000000000000000000000000000000000000000000000076b453ebc0dd7777800000000000000000000000000000000000000000000000002a9ad9cade852ca0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000454f11d58e27858926d7a4ece8bfea2c33e97b13000000000000000000000000c558f600b34a5f69dd2f0d06cb8a88d829b7420a000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc20000000000000000000000005a98fcbea516cf06857215779fd812ca3bef1b32'}


TypeError("Transaction had invalid fields: %r" % invalid)


this will work if I pass it to web3.eth.estimate_gas(tx_prep)    with gas savings

**Comment by fselmo** [comment]
@oocrypticoo I think I see the issue here. The client expects `accessList` as a Tuple of Dicts with `address` and `storageKeys` as keys and their respective values:

```python
access_list = [
    {
        'address':  '0xc558f600b34a5f69dd2f0d06cb8a88d829b7420a',
        'storageKeys': [..., ...]
    },
    {
        'address': '0x454f11d58e27858926d7a4ece8bfea2c33e97b13',
        'storageKeys': [..., ...]
    },
]
```

Whereas `eth-account` (the library web3.py uses to sign off-chain) expects them as Tuples / Lists (no keys, just the values):

```python
access_list = (
    (
        '0xc558f600b34a5f69dd2f0d06cb8a88d829b7420a',
        (
            '0x0000000000000000000000000000000000000000000000000000000000000006',
            '0x0000000000000000000000000000000000000000000000000000000000000007',
        ),
)
```

If you need to sign locally with a private key, give the Tuple setup a try. Let me know if this unblocks you for now, unideal as it may be. We'll look into making this process smoother.

**Comment by oocrypticoo** [comment]
> @oocrypticoo I think I see the issue here. The client expects `accessList` as a Tuple of Dicts with `address` and `storageKeys` as keys and their respective values:
> 
> ```python
> access_list = [
>     {
>         'address':  '0xc558f600b34a5f69dd2f0d06cb8a88d829b7420a',
>         'storageKeys': [..., ...]
>     },
>     {
>         'address': '0x454f11d58e27858926d7a4ece8bfea2c33e97b13',
>         'storageKeys': [..., ...]
>     },
> ]
> ```
> 
> Whereas `eth-account` (the library web3.py uses to sign off-chain) expects them as Tuples / Lists (no keys, just the values):
> 
> ```python
> access_list = (
>     (
>         '0xc558f600b34a5f69dd2f0d06cb8a88d829b7420a',
>         (
>             '0x0000000000000000000000000000000000000000000000000000000000000006',
>             '0x0000000000000000000000000000000000000000000000000000000000000007',
>         ),
>     ),
> )
> ```
> 
> If you need to sign locally with a private key, give the Tuple setup a try. Let me know if this unblocks you for now, un-ideal as it may be. We'll look into making this process smoother.

Thanks for your help,
I saw this once I dug into the account.py file; I figured it was going to be a dict.  That works, excited to see it hit the chain.

**Comment by fselmo** [comment]
Since this isolated issue has been resolved, tracking these improvements in issue #2142
